src/main/java/org/broadinstitute/hellbender/utils/nio/NioFileCopier.java

```diff
@@ -0,0 +1,624 @@
+package org.broadinstitute.hellbender.utils.nio;
+
+import org.apache.commons.collections4.queue.CircularFifoQueue;
+import org.apache.commons.io.FileUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.exceptions.UserException;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Queue;
+
+/**
+ * Class to copy a file using {@link java.nio}.
+ * Operates using paths.
+ *
+ * INSTANCES OF THIS CLASS ARE NOT THREAD-SAFE!
+ *
+ * Created by jonn on 8/27/18.
+ */
+public class NioFileCopier {
```

Comment:
Some general thoughts.  This seems sane and useful.  A good progress meter is the difference between people being patient and people being angry.  

It seems like this class does 3 things, and maybe could be factored differently.  
1.  It downloads a file.
2.  It is a progress meter with a lot of formatting code.
3.  It does some checksum validation.

It seems like maybe the formatting and progress reporting could could be pulled into it's own inner class that would keep track of that state.  
The validation seems like it should be either baked into the download (i.e. as you download each byte[] update the validation and finalize it at the end).  Or just removed from the class and handled externally.  I'd vote for doing it in parallel with the download. 

That will probably end up with more constructor parameters.  It seems like currently this is sort of half a build for configuring this thing, and have a results object.  Maybe they should be split into a downloader / validator factory and then a download result that you can query to get the validation results.  (or have no return type and just throw if it doesn't validate and you asked for validation).

Reply:
Yeah... It does all those things.

I'll make a dummy results object that can be used to query results.  That may make this a little more straight-forward.

The checksum validation has been baked into the download (if requested).